### ISSUE TITLE
Implement Semantic Event Classification System with Bumblebee

### ISSUE THREAD

**Issue by holden** [razrfly/eventasaurus#2843]
## Problem Statement

Our current event categorization system relies on:
1. **YAML-based mappings** - Direct string matching from source categories to internal slugs
2. **Regex/glob patterns** - Pattern matching for known text patterns
3. **Source-specific extractors** - Hardcoded logic for each scraper source

### Current Limitations

- **No semantic understanding**: "live music" and "concert" are not recognized as equivalent
- **Language barriers**: Polish event titles/descriptions require explicit keyword lists
- **Maintenance burden**: Every new pattern requires YAML updates
- **Confidence scores unused**: Always set to 1.0, no probabilistic ranking
- **Poor edge case handling**: Unusual events often fall into "other" category
- **No cross-source learning**: Each source's categorization is independent

### Current System Stats
- 15 predefined categories (with unused hierarchical support)
- 7 source-specific YAML mapping files
- Multiple regex patterns per source
- ~15% of events likely miscategorized or falling back to "other"

## Inspiration: Google Custom Taxonomy Classifier

[Google's Custom Taxonomy Classifier](https://github.com/google-marketing-solutions/custom-taxonomy-classifier) demonstrates a semantic approach:
- Uses **embedding vectors** (mathematical representations of meaning)
- **Vector similarity matching** instead of keyword matching
- Returns **top 10 categories with confidence scores**
- Handles custom taxonomies via spreadsheets

Key insight: Semantic embeddings capture meaning, so "jazz concert" and "live music performance" are recognized as similar without explicit mapping.

## Proposed Solution: Native Elixir with Bumblebee

Instead of a Python microservice, use Elixir's native ML ecosystem:

### Technology Stack
- **[Bumblebee](https://github.com/elixir-nx/bumblebee)** - Pre-trained transformer models in Elixir
- **[Nx](https://github.com/elixir-nx/nx)** - Numerical computing (tensors, linear algebra)
- **[EXLA](https://github.com/elixir-nx/nx/tree/main/exla)** - XLA compiler for acceleration
- **Multilingual model**: `sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2` (supports Polish + 50 languages)

### Why Native Elixir over Python Microservice?
| Factor | Python Microservice | Bumblebee (Native) |
|--------|--------------------|--------------------|
| Latency | HTTP overhead (~50-100ms) | Direct call (~10-20ms) |
| Deployment | Separate service | Same app |
| Maintenance | Two languages | Single codebase |
| Memory | Separate process | Shared BEAM |
| Reliability | Network dependency | No external calls |

### High-Level Flow

```
Event Ingestion
      │
      ▼
┌─────────────────────────────────────────────────────────┐
│                  Classification Pipeline                 │
├─────────────────────────────────────────────────────────┤
│  1. YAML Mapping Lookup (exact match)                   │
│     └── If match found → confidence: 1.0                │
│                                                         │
│  2. Semantic Similarity (Bumblebee)                     │
│     └── Embed event text → cosine similarity → top-K    │
│     └── confidence: 0.5-0.95 based on similarity        │
│                                                         │
│  3. Zero-Shot Classification (fallback)                 │
│     └── BART-large-mnli for ambiguous cases             │
│     └── confidence: 0.3-0.7                             │
└─────────────────────────────────────────────────────────┘
      │
      ▼
Store categories with confidence scores
(public_event_categories join table)
```

#### 1. SemanticClassifier GenServer

```elixir
defmodule EventasaurusDiscovery.Categories.SemanticClassifier do
  use GenServer
  
  # Loads model and pre-computes category embeddings on startup
  def init(_) do
    {:ok, model} = Bumblebee.load_model({:hf, "sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2"})
    {:ok, tokenizer} = Bumblebee.load_tokenizer({:hf, "sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2"})
    
    # Pre-compute category embeddings
    categories = Categories.list_categories()
    category_embeddings = compute_category_embeddings(model, tokenizer, categories)
    
    {:ok, %{model: model, tokenizer: tokenizer, category_embeddings: category_embeddings}}
  end
  
  # Classify event text, return top-K categories with scores
  def classify(text, opts \\ []) do
    GenServer.call(__MODULE__, {:classify, text, opts})
  end
end
```

#### 2. Enhanced Category Schema

```elixir
# Add to categories table
alter table(:categories) do
  add :synonyms, {:array, :string}, default: []      # ["live music", "gig", "show"]
  add :keywords, {:array, :string}, default: []       # ["band", "venue", "ticket"]
  add :embedding_text, :text                          # Combined text for embedding
  add :embedding_vector, :binary                      # Pre-computed embedding (optional cache)
end
```

#### 3. Hybrid Classification Pipeline

```elixir
defmodule EventasaurusDiscovery.Categories.HybridClassifier do
  @doc """
  Classify event using hybrid approach:
  1. YAML exact match (highest confidence)
  2. Semantic similarity (medium confidence)
  3. Zero-shot classification (lower confidence)
  """
  def classify_event(event, source) do
    text = build_event_text(event)
    
    # Try YAML mapping first
    case YamlMapper.classify(text, source) do
      {:ok, categories} when categories != [] ->
        {:ok, add_confidence(categories, 1.0)}
      
      _ ->
        # Fall back to semantic classification
        semantic_classify(text)
    end
  end
  
  defp semantic_classify(text) do
    case SemanticClassifier.classify(text, top_k: 3, min_confidence: 0.5) do
      {:ok, results} -> {:ok, results}
      {:error, _} -> {:ok, [{:other, 0.3}]}  # Fallback to "other"
    end
  end
end
```

### Phase 1: Foundation (Week 1-2)
- [ ] Add Bumblebee, Nx, EXLA dependencies
- [ ] Create SemanticClassifier GenServer
- [ ] Load multilingual embedding model
- [ ] Implement basic text → embedding → similarity pipeline
- [ ] Write tests with sample events

### Phase 2: Category Taxonomy Enhancement (Week 2-3)
- [ ] Add synonyms, keywords fields to Category schema
- [ ] Populate synonyms for all 15 categories
- [ ] Add Polish translations for category descriptions
- [ ] Create embedding_text field (combined category info)
- [ ] Pre-compute and cache category embeddings

### Phase 3: Classification Pipeline (Week 3-4)
- [ ] Implement HybridClassifier module
- [ ] Integrate with existing CategoryExtractor
- [ ] Add meaningful confidence scores
- [ ] Implement multi-category assignment with primary selection
- [ ] Add monitoring/metrics for classification accuracy

### Phase 4: Integration & Refinement (Week 4-5)
- [ ] Integrate into scraper pipeline
- [ ] Add admin UI for viewing/correcting classifications
- [ ] Build evaluation dataset from correctly classified events
- [ ] Tune confidence thresholds
- [ ] Document the system

## Acceptance Criteria

- [ ] Events are classified with meaningful confidence scores (0.0-1.0)
- [ ] Semantic similarity works across languages (English + Polish)
- [ ] Classification latency < 50ms per event
- [ ] Graceful degradation when model unavailable
- [ ] "Other" category usage reduced by 50%+
- [ ] Admin can view and correct classifications
- [ ] Monitoring shows classification distribution

### Memory & Resources
- Model size: ~500MB-1GB in memory
- Startup time: 5-10 seconds for model loading
- Can use Fly.io GPU instances if needed for EXLA acceleration

### Multilingual Support
- `paraphrase-multilingual-MiniLM-L12-v2` supports 50+ languages including Polish
- Alternative: `intfloat/multilingual-e5-base` for higher accuracy

### Caching Strategy
- Pre-compute category embeddings on startup
- Cache frequently seen event text patterns
- ETS table for embedding cache

### Graceful Degradation
- If Bumblebee fails to load → fall back to YAML-only classification
- If classification times out → use "other" with low confidence
- Log failures for debugging

## Open Questions

1. **Model selection**: Should we start with MiniLM (faster, smaller) or E5 (more accurate)?
2. **Subcategories**: Should we implement hierarchical categories (Concerts → Rock, Jazz, Classical)?
3. **Zero-shot**: Is zero-shot classification worth the complexity, or is embedding similarity sufficient?
4. **Training data**: Should we build a manually-labeled dataset for evaluation/fine-tuning?
5. **GPU acceleration**: Is EXLA/GPU worth the infrastructure complexity?

## References

- [Google Custom Taxonomy Classifier](https://github.com/google-marketing-solutions/custom-taxonomy-classifier)
- [Bumblebee Documentation](https://hexdocs.pm/bumblebee)
- [Sentence Transformers](https://www.sbert.net/)
- [DockYard: Semantic Search with Phoenix, Axon, Bumblebee](https://dockyard.com/blog/2023/01/11/semantic-search-with-phoenix-axon-bumblebee-and-exfaiss)
- [Elixir Merge: Using Nomic Embed with Bumblebee](https://elixirmerge.com/p/using-nomic-embed-with-elixir-bumblebee-for-text-embeddings)
- [Text Embeddings, Classification, and Semantic Search](https://towardsdatascience.com/text-embeddings-classification-and-semantic-search-8291746220be/)

## Related Files

Current categorization system:
- \`lib/eventasaurus_discovery/categories/category_extractor.ex\`
- \`lib/eventasaurus_discovery/categories/category_mapper.ex\`
- \`priv/category_mappings/*.yml\`
- \`priv/repo/seeds/reference_data/categories.exs\`
